File: FTP Client for WinForms using Ultimate FTP library/[C#]-FTP Client for WinForms using Ultimate FTP library/C#/ClientController_Download.cs
using System;
using System.Diagnostics;
using System.Drawing;
using System.IO;
using ComponentPro;
using ComponentPro.IO;

namespace ClientSample
{
    partial class ClientController
    {
        public void DoRemoteDownload(AbstractFileInfo[] files)
        {
            if (_state != ConnectionState.Ready)
                return;

            EnableProgress(true);

            bool showProgress = _settings.Get<bool>(SettingInfo.ShowProgressWhileTransferring);

            TransferOptions opt = new TransferOptions(showProgress, RecursionMode.RecurseIntoAllSubFolders, true,
                (SearchCondition)null, FileExistsResolveAction.Confirm, SymlinksResolveAction.Confirm);

#if Framework4_5 || !ASYNC
            FileSystemTransferStatistics stat;
            try
            {
#if Framework4_5
                stat = await client.DownloadFilesAsync(_currentDirectory, fileList, _currentLocalDirectory, opt);
#else
                stat = client.DownloadFiles(_currentDirectory, fileList, _currentLocalDirectory, opt);
#endif
                ShowStatistics(stat, "download");
            }
            catch (Exception ex)
            {
                if (!HandleException(ex))
                    return;
            }            
            
            EnableProgress(false);
            RefreshLocalList();
#else
            _client.DownloadFilesAsync(_currentDirectory, files, _currentLocalDirectory, opt);
#endif
        }

        /// <summary>
        /// Handles the client's DownloadFilesCompleted event.
        /// </summary>
        /// <param name="sender">The client object.</param>
        /// <param name="e">The event arguments.</param>
        void client_DownloadFilesCompleted(object sender, ExtendedAsyncCompletedEventArgs<FileSystemTransferStatistics> e)
        {
            if (e.Error != null)
            {
                if (!HandleException(e.Error))
                    return;
            }
            else
                ShowStatistics(e.Result, "download");

            EnableProgress(false);
            RefreshLocalList();
        }

        public void DoRemoteResumeDownload(AbstractFileInfo[] files)
        {
            if (files == null)
                return;

            EnableProgress(true);
            try
            {
                for (int i = 0; i < files.Length; i++)
                {
                    if (_aborting)
                        break;

                    AbstractFileInfo remoteFile = files[i];

                    if (remoteFile.IsFile)
                    {
                        // Download a single file.
                        long result = _client.ResumeDownloadFile(remoteFile.Name, Path.Combine(_currentLocalDirectory, remoteFile.Name));

                        if (files.Length == 1)
                        {
                            if (result == -1)
                                _view.ShowMessage("Local file size is greater than the remote file size");
                            else if (result == 0)
                                _view.ShowMessage("Local file size is equal to the remote file size");
                        }
                    }
                }
            }
            catch (Exception exc)
            {
                if (!HandleException(exc))
                    return;
            }

            EnableProgress(false);
            RefreshLocalList();
        }

        public void DoRemoteView(string name)
        {
            string tempFile = Util.GetTempFileName(name);
            string remoteFile = name;// GetRemoteFullPath(item.Text);

            EnableProgress(true);

            _client.DownloadFileCompleted += client_GetFileForViewingCompleted;
            _client.DownloadFileAsync(remoteFile, tempFile, tempFile);
        }

        /// <summary>
        /// Handles the client's DownloadFileCompleted event.
        /// </summary>
        /// <param name="sender">The Ftp object.</param>
        /// <param name="e">The event arguments.</param>
        private void client_GetFileForViewingCompleted(object sender, ExtendedAsyncCompletedEventArgs<long> e)
        {
            // Detach the event handler.
            _client.DownloadFileCompleted -= client_GetFileForViewingCompleted;

            bool aborted = false;

            if (e.Error != null)
            {
                if (!HandleException(e.Error))
                    return;

                aborted = Util.IsFileSystemOperationCancelled(e.Error);
            }

            // temporary file name is saved in the async state object.
            string tempFile = (string)e.UserState;

            DoRemoteViewPost(aborted, tempFile);
        }

        void DoRemoteViewPost(bool aborted, string tempFile)
        {
            EnableProgress(false);

            if (_state == ConnectionState.Disconnecting)
            {
                Disconnect();
                return;
            }

            if (!aborted)
            {
                try
                {
                    Process.Start("Notepad.exe", tempFile);
                }
                catch (Exception exc)
                {
                    _view.ShowError("Error:" + exc.Message);
                }
            }
        }
    }
}

File: How to write an SFTP file transfer client for WinForms/[VB]-How to write an SFTP file transfer client for WinForms/VB/ClientController_Download.vb
Imports System.IO
Imports ComponentPro
Imports ComponentPro.IO

Namespace ClientSample
	Partial Friend Class ClientController
		''' <summary>
		''' Downloads remote files.
		''' </summary>
		''' <param name="files">Files to download.</param>
		Public Sub DoRemoteDownload(ByVal files() As AbstractFileInfo)
			If _state <> ConnectionState.Ready Then
				Return
			End If

			EnableProgress(True)

			Dim showProgress As Boolean = _settings.Get(Of Boolean)(SettingInfo.ShowProgressWhileTransferring)

			Dim opt As New TransferOptions(showProgress, RecursionMode.RecurseIntoAllSubFolders, True, CType(Nothing, SearchCondition), FileExistsResolveAction.Confirm, SymlinksResolveAction.Confirm)

#If Framework4_5 OrElse (Not ASYNC) Then
			Dim stat As FileSystemTransferStatistics
			Try
#If Framework4_5 Then
				stat = Await client.DownloadFilesAsync(_currentDirectory, fileList, _currentLocalDirectory, opt)
#Else
				stat = client.DownloadFiles(_currentDirectory, fileList, _currentLocalDirectory, opt)
#End If
				ShowStatistics(stat, "download")
			Catch ex As Exception
				If Not HandleException(ex) Then
					Return
				End If
			End Try

			EnableProgress(False)
			RefreshLocalList()
#Else
			_client.DownloadFilesAsync(_currentDirectory, files, _currentLocalDirectory, opt)
#End If
		End Sub

		''' <summary>
		''' Handles the client's DownloadFilesCompleted event.
		''' </summary>
		''' <param name="sender">The client object.</param>
		''' <param name="e">The event arguments.</param>
		Private Sub client_DownloadFilesCompleted(ByVal sender As Object, ByVal e As ExtendedAsyncCompletedEventArgs(Of FileSystemTransferStatistics))
			If e.Error IsNot Nothing Then
				If Not HandleException(e.Error) Then
					Return
				End If
			Else
				ShowStatistics(e.Result, "download")
			End If

			EnableProgress(False)
			RefreshLocalList()
		End Sub

		''' <summary>
		''' Resumes download the specified files.
		''' </summary>
		''' <param name="files">Files to resume download.</param>
		Public Sub DoRemoteResumeDownload(ByVal files() As AbstractFileInfo)
			If files Is Nothing Then
				Return
			End If

			EnableProgress(True)
			Try
				For i As Integer = 0 To files.Length - 1
					If _aborting Then
						Exit For
					End If

					Dim remoteFile As AbstractFileInfo = files(i)

					If remoteFile.IsFile Then
						' Download a single file.
						Dim result As Long = _client.ResumeDownloadFile(remoteFile.Name, Path.Combine(_currentLocalDirectory, remoteFile.Name))

						If files.Length = 1 Then
							If result = -1 Then
								_view.ShowMessage("Local file size is greater than the remote file size")
							ElseIf result = 0 Then
								_view.ShowMessage("Local file size is equal to the remote file size")
							End If
						End If
					End If
				Next i
			Catch exc As Exception
				If Not HandleException(exc) Then
					Return
				End If
			End Try

			EnableProgress(False)
			RefreshLocalList()
		End Sub

		''' <summary>
		''' Views the remote file.
		''' </summary>
		''' <param name="name">File to view.</param>
		Public Sub DoRemoteView(ByVal name As String)
			Dim tempFile As String = Util.GetTempFileName(name)
			Dim remoteFile As String = name ' GetRemoteFullPath(item.Text);

			EnableProgress(True)

			AddHandler _client.DownloadFileCompleted, AddressOf client_GetFileForViewingCompleted
			_client.DownloadFileAsync(remoteFile, tempFile, tempFile)
		End Sub

		''' <summary>
		''' Handles the client's DownloadFileCompleted event.
		''' </summary>
		''' <param name="sender">The Ftp object.</param>
		''' <param name="e">The event arguments.</param>
		Private Sub client_GetFileForViewingCompleted(ByVal sender As Object, ByVal e As ExtendedAsyncCompletedEventArgs(Of Long))
			' Detach the event handler.
			RemoveHandler _client.DownloadFileCompleted, AddressOf client_GetFileForViewingCompleted

			Dim aborted As Boolean = False

			If e.Error IsNot Nothing Then
				If Not HandleException(e.Error) Then
					Return
				End If

				aborted = Util.IsFileSystemOperationCancelled(e.Error)
			End If

			' temporary file name is saved in the Async state object.
			Dim tempFile As String = CStr(e.UserState)

			DoRemoteViewPost(aborted, tempFile)
		End Sub

		Private Sub DoRemoteViewPost(ByVal aborted As Boolean, ByVal tempFile As String)
			EnableProgress(False)

			If _state = ConnectionState.Disconnecting Then
				Disconnect()
				Return
			End If

			If Not aborted Then
				Try
					Process.Start("Notepad.exe", tempFile)
				Catch exc As Exception
					_view.ShowError("Error:" & exc.Message)
				End Try
			End If
		End Sub
	End Class
End Namespace
